Login implement

Lighty0410 opened the pull request:
1. Add login 
2. Implement password security

Lighty0410 committed:
I should check this at home

Study test cases

Lighty0410 committed:
Official mongo implementation

Lighty0410 committed:
Server func login implementation

Lighty0410 committed:
Implement login, password security, add commentaries

Lighty0410 committed:
Minor fixes in user/crypto/register

Lighty0410 committed:
Merge branch 'master' into password_implementation

Lighty0410 committed:
Add different code statuses

Lighty0410 committed:
Compare and return different status codes

Lighty0410 committed:
Fix status code error.

Homework

sashayakovtseva commented:
I suggest renaming this PR to smth like 'Implement login'. How can you implement a password?

sashayakovtseva commented:
Do not end commit messages with a dot

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -14,7 +18,8 @@ type User struct {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
 	c := s.db.Collection("users")
 	_, err := c.InsertOne(ctx, u) // err := c.Insert(u)
```
remove debug comments, whey are cluttering code

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -23,13 +28,20 @@ func (s *Service) AddUser(u *User) error {
+	var result User
+	filter := bson.D{{"name", username}}
+	err := s.db.Collection("users").FindOne(context.Background(), filter).Decode(&result)
+	fmt.Println(result.Name, result.Hash)
```
why you print here?

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -23,13 +28,20 @@ func (s *Service) AddUser(u *User) error {
+	filter := bson.D{{"name", username}}
+	err := s.db.Collection("users").FindOne(context.Background(), filter).Decode(&result)
+	fmt.Println(result.Name, result.Hash)
+	switch err != nil {
```
although it seems to be correct syntax, this looks really weird. More common and easy to read would be `switch err`

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -23,13 +28,20 @@ func (s *Service) AddUser(u *User) error {
+	err := s.db.Collection("users").FindOne(context.Background(), filter).Decode(&result)
+	fmt.Println(result.Name, result.Hash)
+	switch err != nil {
+	case err == errors.New("mongo: no documents in result"):
```
this will not work. also check Decode documentation for better comparison here

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -23,13 +28,20 @@ func (s *Service) AddUser(u *User) error {
+	switch err != nil {
+	case err == errors.New("mongo: no documents in result"):
+		return result, http.StatusUnauthorized, err
+	case err == errors.New("Registry cannot be nil"):
```
as I said, no need to check this error at all

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -23,13 +28,20 @@ func (s *Service) AddUser(u *User) error {
+	case err == errors.New("Registry cannot be nil"):
+		return result, http.StatusUnauthorized, err
+	case err != nil:
+		return result, http.StatusInternalServerError, err
```
Why mongo package knows anything about HTTP status codes? this is a bad design. db package should operate on custom errors only and do not rely on other top-level packages

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -23,13 +28,20 @@ func (s *Service) AddUser(u *User) error {
-	if err != nil {
-		log.Println("incorrect collection")
+// ReadUser retrieves an information from the database and compares it with a request.
+func (s *Service) ReadUser(username string) (User, int, error) {
```
Tree return values - too much!

sashayakovtseva commented on pkg/server/crypto.go:
```diff
@@ -0,0 +1,19 @@
+func generateHash(password string) (string, error) {
+	cryptoHash, err := bcrypt.GenerateFromPassword([]byte(password), 15)
+	if err != nil {
+		log.Println("Incorrect password") // Are there any requirements to return string/error in this case ?
```
remove debug comments.
and yes, you should return error here

sashayakovtseva commented on pkg/server/crypto.go:
```diff
@@ -0,0 +1,19 @@
+	if err != nil {
+		log.Println("Incorrect password") // Are there any requirements to return string/error in this case ?
+	}
+	return string(cryptoHash), err
```
`return string(cryptoHash), nil` is clear and easy to read

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -3,27 +3,39 @@ package server
 // handleRegistration registers user in the system.
 func (s *EkadashiServer) handleRegistration(w http.ResponseWriter, r *http.Request) {
 	var req registerRequest
+	var password string
```
no need for var, use short syntax

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -3,27 +3,39 @@ package server
+	}
+	password, err = generateHash(req.Password)
+	if err != nil {
+		log.Println("Incorrect password")
```
return correct error here so that user can understand it

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -3,27 +3,39 @@ package server
+		log.Println("Incorrect password")
 		return
 	}
+	fmt.Println(req.Username)
```
why you print to stdout?

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -3,27 +3,39 @@ package server
 	err = s.db.AddUser(&mongo.User{
 		Name: req.Username,
-		Hash: req.Password, //hash
+		Hash: password,
```
a different name for password will be better

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -32,6 +44,25 @@ func (s *EkadashiServer) handleRegistration(w http.ResponseWriter, r *http.Reque
+	var req loginRequest
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		jsonError(w, http.StatusBadRequest, fmt.Errorf("can not decode the request.:  %v", err))
```
do not write dots at the end

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -32,6 +44,25 @@ func (s *EkadashiServer) handleRegistration(w http.ResponseWriter, r *http.Reque
+	}
+	user, status, err := s.db.ReadUser(req.Username)
+	if err != nil {
+		jsonError(w, status, fmt.Errorf("pswd: %v", err))
```
what does pswd mean?

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -32,6 +44,25 @@ func (s *EkadashiServer) handleRegistration(w http.ResponseWriter, r *http.Reque
+	if err != nil {
+		jsonError(w, status, fmt.Errorf("pswd: %v", err))
+		return
+	} else if status != http.StatusUnauthorized && status != http.StatusOK && err != nil {
```
no need to use `else if` when you return before

sashayakovtseva commented on pkg/server/register.go:
```diff
@@ -32,6 +44,25 @@ func (s *EkadashiServer) handleRegistration(w http.ResponseWriter, r *http.Reque
+	}
+	err = compareHash(user.Hash, []byte(req.Password))
+	if err != nil {
+		jsonError(w, http.StatusForbidden, fmt.Errorf("incorrect username or password: %v", err))
```
401, not 403

sashayakovtseva commented:
Linter errors: 
```
pkg/mongo/user.go:34::error: github.com/mongodb/mongo-go-driver/bson/primitive.E composite literal uses unkeyed fields (vet)
pkg/mongo/user.go:40:24:warning: error strings should not be capitalized (ST1005) (staticcheck)
make: *** [lint] Error 1
```

sashayakovtseva commented:
Also missing test for `ReadUser`

Lighty0410 committed:
Redesign user login

sashayakovtseva commented on pkg/mongo/user.go:
```diff
@@ -12,24 +15,28 @@ type User struct {
+	var hash User
+	filter := bson.D{{Key: "name", Value: username}}
+	err := s.db.Collection("users").FindOne(context.Background(), filter).Decode(&hash)
+	if err == mongo.ErrNoDocuments {
```
What if `err != mongo.ErrNoDocuments`?

sashayakovtseva commented:
Still no test for `ReadUser`

Lighty0410 committed:
Add test for login functionality

Lighty0410 committed:
Add test for login functionality

Lighty0410 committed:
Implement better solution for database tests

Lighty0410 committed:
Fix an empty string issue in the databasetest 

Actually that's not a proper resolution of this problem until we'll set a validation

Lighty0410 committed:
Fix an empty string issue in the databasetest 

Actually that's not a proper resolution of this problem until we'll set a validation

Lighty0410 committed:
Merge remote-tracking branch 'origin/password_implementation' into password_implementation

Lighty0410 committed:
Final fix
